### ISSUE TITLE
API proposal - Chaining commands that use indexes

### ISSUE THREAD

**Issue by neumino** [rethinkdb/rethinkdb#2670]
Slightly related to https://github.com/rethinkdb/rethinkdb/issues/955 and the comment https://github.com/rethinkdb/rethinkdb/issues/2571#issuecomment-47030366

Currently:
- We can chain `between` and `orderBy` if they use the same index
- We can chain `between` and `group` if they use the same index

While we cannot efficiently intersect indexes, it should be possible to have slightly better performance when chaining commands that use partially a compound index

Suppose the compound index is built in `field1` and `field2`.

```
r.table("test").getAll(<k values for field1>, {partial: "compound"}).orderBy({partial: "compound"})
```

Where `partial` means "a part of the compound index". This should be somehow doable using a heap in `O(k)+O(log(k)*n)` where `n` is the total number of document to return.

The same thing applies for chaining `between` and `orderBy`.

It should also be possible to have a relatively efficient algorithm to chain `getAll` and `group` without blowing the memory.

I'm putting it in backlog for now (I first want to be able to point to an issue when users hit this problem)

**Comment by jwr** [comment]
I think I've just hit this limitation, so here's a use case for @neumino:

I have a table that contains data with (among others) {:owner uid :last-op timestamp}. I often need to get all documents that belong to a particular owner, sorted by timestamp. I have a compound index on [uid timestamp]. I was certain I could just use a getAll/orderBy with this compound index, but it turned out I could not. I then tried using two separate indices (one on :owner and one on :last-op), but this only got me to "Indexed order_by can only be performed on a TABLE or TABLE_SLICE".

From what I understand, the only option right now is to get all documents for a given uid using an index on :owner and then do an in-memory sort. This will not work well for larger amounts of documents.

**Comment by thelinuxlich** [comment]
can't you use between + orderBy?

**Comment by jwr** [comment]
Perhaps I can, but I don't know how. I tried, but I get "Cannot order by index `last-op` after calling BETWEEN on index `owner`" (I just tried doing (r/between 1 1 {:index "owner"}) (r/order-by {:index (r/desc "last-op")}).

From the docs:

> Currently, you cannot chain an orderBy using a secondary index after a getAll but only after a table. (You can, however, chain an orderBy using a secondary index after a between if they use the same index.)

So I guess I can't, because those would be two different indices.

**Comment by thelinuxlich** [comment]
Yeah, right now it works only if it is the same index for between and orderBy, can't you use a compound index with owner and last_op?

**Comment by coffeemug** [comment]
Currently you'd have to do an indexed `between` and chain on `orderBy` without an index. I'm not actually sure if other databases support this type of behavior -- doing a query by one index and ordering by another index is algorithmically very hard, and it's not clear to me that it's actually implementable in a way that's more efficient than a memory sort.

We will look into it, though.

**Comment by jwr** [comment]
I tried doing:

    (r/between [1 1] [0 r/maxval] {:index "owner-last-op"})
    (r/order-by {:index (r/desc "owner-last-op")})

using my compound index on [owner last-op]. It seems to work, at least it doesn't throw an error, and returns the right results. I don't know what it does on the server side.

@coffeemug I think this shouldn't be a problem if we're dealing with a single compound index?

**Comment by thelinuxlich** [comment]
yeah, that's the way I'd do it :)

**Comment by coffeemug** [comment]
Oh sorry, that would work. However, if you were querying multiple owners, this query would sort by owners first, and break ties by `last-op`.

**Comment by jwr** [comment]
Well, it works for me. The only disadvantage is the syntax: you have to know that this is exactly how one needs to approach it. What @neumino proposed initially looks better.

**Comment by mike-marcacci** [comment]
I've been following this feature for quite some time, and it's got to be on the very top of my wish-list (along with multi-document atomic updates). While the real solution is to make intersections efficient, this *might* be able to be faked in the meantime (with permanently useful semantics):

RethinkDB already checks for situations where multiple indices are used (and throws an error). It also is already aware of how each query is built – the simple field, the computed function, etc. It seems to me that it would be reasonably easy to use the first index, then apply the other indices' *rules* to the result set.

Of course, this has "correctness" implications (you're telling it to use an index and it isn't) but until intersections are possible, we already have to pick **one** index, then filter/sort the result set. To fix this (and provide a really cool third option to the whole auto-optimization debate) we could add an option to the tables and the `run` function that enables index downgrading behavior. Because there are times when you wouldn't want to use a secondary/tertiary index even if you had one capable of intersections (ie. the result set is very small) index downgrading is a form of query optimization that is very predictable and doesn't require the database to pick between simple and computed indices because they are user-supplied. This also solves many of the use cases that are listed in #2356 and #4150: the user can *always* specify which query to use, and not worry about how the intersection works internally.

Anyhow, just wanted to throw this out there because building an auto-optimizer that has to pick between several computed indices sounds like quite a chore, and this might be an acceptable solution to a number of issues here!

**Comment by danielmewes** [comment]
Thanks for your comment @mike-marcacci . We actually discussed some very similar ideas in https://github.com/rethinkdb/rethinkdb/issues/4150 as you might have seen.

**Comment by mike-marcacci** [comment]
Ah yes, I had missed the specifics of your suggestion in [this comment](https://github.com/rethinkdb/rethinkdb/issues/4150#issuecomment-133204462). You can add my enthusiastic :+1: to that proposal!
